src/functions/InModuleScope.ps1

```diff
@@ -75,26 +75,48 @@
     )
 
     $module = Get-ScriptModule -ModuleName $ModuleName -ErrorAction Stop
+    $sessionState = Set-SessionStateHint -PassThru -Hint "Module - $($module.Name)" -SessionState $module.SessionState
 
-    # TODO: could this simply be $PSCmdlet.SessionState? Because the original scope we are moving from
-    # is the scope in which this command is running, right?
-    # $originalState = $Pester.SessionState
-    # $originalScriptBlockScope = Get-ScriptBlockScope -ScriptBlock $ScriptBlock
-
-    # try {
-    # $sessionState = Set-SessionStateHint -PassThru -Hint "Module - $($module.Name)" -SessionState $module.SessionState
-    # $Pester.SessionState = $sessionState
-    # Set-ScriptBlockScope -ScriptBlock $ScriptBlock -SessionState $sessionState
-
-    # do {
-    # Write-ScriptBlockInvocationHint -Hint "InModuleScope" -ScriptBlock $ScriptBlock
-    & $module $ScriptBlock @Parameters @ArgumentList
-    # } until ($true)
-    # }
-    # finally {
-    # $Pester.SessionState = $originalState
-    # Set-ScriptBlockScope -ScriptBlock $ScriptBlock -SessionStateInternal $originalScriptBlockScope
-    # }
+    $wrapper = {
+        param (
+            [Parameter(Mandatory = $true)]
+            [scriptblock]
+            ${Script Block},
+
+            [hashtable]
+            $___Parameters___ = @{ },
+
+            [object[]]
+            $___ArgumentList___ = @(),
+
+            [System.Management.Automation.SessionState]
+            ${Session State},
+
+            ${Set Dynamic Parameter Variable}
+        )
+
+        # This script block is used to create variables for provided parameters that
+        # the real scriptblock can inherit. Makes defining a param-block optional.
+
+        & ${Set Dynamic Parameter Variable} -SessionState ${Session State} -Parameters $___Parameters___
+
+        # define this in the current scope to be used instead of $PSBoundParameter if needed
+        # $PesterBoundParameters = if ($null -ne $___Parameters___) { $___Parameters___ } else { @{} }
+        & ${Script Block} @___Parameters___ @___ArgumentList___
+    }
+
+    Set-ScriptBlockScope -ScriptBlock $ScriptBlock -SessionState $sessionState
+    Set-ScriptBlockScope -ScriptBlock $wrapper -SessionState $sessionState
+    $splat = @{
+        'Script Block'                   = $ScriptBlock
+        '___ArgumentList___'             = $ArgumentList
+        '___Parameters___'               = $Parameters
+        'Session State'                  = $sessionState
+        'Set Dynamic Parameter Variable' = $SafeCommands['Set-DynamicParameterVariable']
+    }
+
+    Write-ScriptBlockInvocationHint -Hint "InModuleScope" -ScriptBlock $ScriptBlock
```

Comment:
The hint looks okay. 

I added them for one case and at that time it seemed like a great idea, but I don't remember ever using it again. It is sometimes helpful to have an Id on the session state, but normally I just check the $executioncontext.sessionsstate and look at module to see in which session state I am in, and when debugging issues with having multiple session states with the same name it is usually together with Mock, so I just look for Get-Command Pester* which gives me mocks with id's and I can tell if I am in the same session state. 

I actually plan to get rid of the hints. Or maybe we should incorporate them into the diag log? I need to see how verbose it is and how much I can gather from them, e.g. when debugging an issue like the one where scriptblock was running in a different session state than the wrapper.